taskcat/s3_sync.py

```diff
@@ -162,14 +164,15 @@ def _s3_upload_file(self, paths, prefix, s3_client, acl):
         retry = 0
         # backoff and retry
         while retry < 5:
+            print("{}[S3: -> ]{} s3://{}/{}".format(PrintMsg.white, PrintMsg.rst_color, bucket, prefix + s3_path))
             try:
                 s3_client.upload_file(local_filename, bucket, prefix + s3_path, ExtraArgs={'ACL': acl})
-                print("{}[S3: -> ]{} s3://{}/{}".format(PrintMsg.white, PrintMsg.rst_color, bucket, prefix + s3_path))
                 break
             except Exception as e:
                 retry += 1
-                if retry == 5:
-                    raise
-                time.sleep(retry * 2)
                 print(PrintMsg.ERROR + "S3 upload error: %s" % e)
+                # give up if we've exhausted retries, or if the error is not-retryable (ie AccessDenied)
+                if retry == 5 or (type(e) == S3UploadFailedError and '(AccessDenied)' in str(e)):
```

Comment:
`s3_client.exceptions.S3UploadFailedError` instead of `S3UploadFailedError`? 

Is there a better way to catch `'(AccessDenied)' `versus the 'in-string' method?

Reply:
not that I can see 😞

Reply:
only properties the error contains is `args` which only contains the string. No `Code` or anything else...

```
In [15]: type(ex) == S3UploadFailedError
Out[15]: True

In [16]: dir(ex)
Out[16]:
['__cause__',
 '__class__',
 '__context__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__suppress_context__',
 '__traceback__',
 '__weakref__',
 'args',
 'with_traceback']

In [17]: dir(ex.args)
Out[17]:
['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'count',
 'index']

In [18]: ex.args
Out[18]: ('Failed to upload /Users/jmmccon//broker.yaml to aws-quickstart/test: An error occurred (AccessDenied) when calling the PutObject operation: Access Denied',)
```